### ISSUE TITLE
Haproxy 2.x openssl/libressl issues with curves bind option (Opnsense)

### ISSUE THREAD

**Issue by waffshappen** [haproxy/haproxy#366]
This is coming from a puzzling issue over at:

https://github.com/opnsense/plugins/issues/1573

## Output of `haproxy -vv` and `uname -a`

```
HA-Proxy version 2.0.8 2019/10/23 - https://haproxy.org/
Build options :
  TARGET  = freebsd
  CPU     = generic
  CC      = cc
  CFLAGS  = -O2 -pipe -DHARDENEDBSD -fPIE -fPIC -fstack-protector-all -fno-strict-aliasing -fno-strict-aliasing -Wdeclaration-after-statement -fwrapv -Wno-address-of-packed-member -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-ignored-qualifiers -Wno-missing-field-initializers -Wno-implicit-fallthrough -Wtype-limits -Wshift-negative-value -Wnull-dereference -DFREEBSD_PORTS
  OPTIONS = USE_PCRE=1 USE_PCRE_JIT=1 USE_REGPARM=1 USE_STATIC_PCRE=1 USE_GETADDRINFO=1 USE_OPENSSL=1 USE_LUA=1 USE_ACCEPT4=1 USE_ZLIB=1 USE_CPU_AFFINITY=1

Feature list : -EPOLL +KQUEUE -MY_EPOLL -MY_SPLICE -NETFILTER +PCRE +PCRE_JIT -PCRE2 -PCRE2_JIT +POLL -PRIVATE_CACHE +THREAD -PTHREAD_PSHARED +REGPARM +STATIC_PCRE -STATIC_PCRE2 +TPROXY -LINUX_TPROXY -LINUX_SPLICE +LIBCRYPT -CRYPT_H -VSYSCALL +GETADDRINFO +OPENSSL +LUA -FUTEX +ACCEPT4 -MY_ACCEPT4 +ZLIB -SLZ +CPU_AFFINITY -TFO -NS -DL -RT -DEVICEATLAS -51DEGREES -WURFL -SYSTEMD -OBSOLETE_LINKER -PRCTL -THREAD_DUMP -EVPORTS

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=2).
Built with OpenSSL version : LibreSSL 2.9.2
Running on OpenSSL version : LibreSSL 2.9.2
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2
Built with Lua version : Lua 5.3.5
Built with transparent proxy support using: IP_BINDANY IPV6_BINDANY
Built with zlib version : 1.2.11
Running on zlib version : 1.2.11
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with PCRE version : 8.43 2019-02-23
Running on PCRE version : 8.43 2019-02-23
PCRE library supports JIT : yes
Encrypted password support via crypt(3): yes

Available polling systems :
     kqueue : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use kqueue.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
              h2 : mode=HTTP       side=FE        mux=H2
              h2 : mode=HTX        side=FE|BE     mux=H2
       <default> : mode=HTX        side=FE|BE     mux=H1
       <default> : mode=TCP|HTTP   side=FE|BE     mux=PASS

Available services : none

Available filters :
	[SPOE] spoe
	[COMP] compression
	[CACHE] cache
	[TRACE] trace

```
```
FreeBSD [hostname] 11.2-RELEASE-p14-HBSD FreeBSD 11.2-RELEASE-p14-HBSD  07680caafe9(stable/19.7)  amd64
```

## What's the configuration?

```
Auto-generated by the opnsense haproxy plugin, it concerns only the "curves" bind option, everything else works
```

## Steps to reproduce the behavior

1. Setup opnsense 19.7.6 with libressl and the haproxy plugin
2. Enable "advanced" and put "curves X25519" inside the bind pass-through options
3. Enable haproxy

## Actual behavior

Haporxy fails to start, a config check reveals neither the built openssl or libressl support "X25519" as curve - and libressl fails even with the much more basic sect/sept curves.

## Expected behavior

Haproxy to start and prefer the *25519 curves as first option, and have it available

## Do you have any idea what may have caused this?
(Likely) the update to haproxy 2.x from opnsense

## Do you have an idea how to solve the issue?
Fixing whatever keeps openssl to work with X25519 ( https://cbonte.github.io/haproxy-dconv/2.0/configuration.html#5.1-curves ) options straight from the docs, and libressl to fail with ANY "curves" option specified.

I dont know if this is an issue in the build of opnsense, a haproxy bug, or if the docs need to be changed - it definitely has an issue with libressl as not even `curves secp384r1` work, with the error that the library doesnt support the suite, yet it being listed in openssl ec/ecparam, while openssl doesnt work with X25519 specified - and even if it would it doesnt handle `TLS_CHACHA20_POLY1305_SHA256` on opnsense for me as cipher.

**Comment by wtarreau** [comment]
Without at least a piece of configuration exhibiting the problem it will be difficult to make any progress on this I'm afraid. Maybe just make a copy of your generated config and trim it to only keep the parts causing the failure ?

**Comment by waffshappen** [comment]
> Without at least a piece of configuration exhibiting the problem it will be difficult to make any progress on this I'm afraid. Maybe just make a copy of your generated config and trim it to only keep the parts causing the failure ?

While i dont think the entire config is needed - or even bits of it i will provide it below.

Because the important part of that config is, as mentioned, the "curves" bind option:

```
frontend HTTPS_PUB
    bind internalip:443 name internalip:443 curves X25519:secp521r1:secp384r1
```

Overview:
pre-haproxy 2.0 opnsense
openssl: works
libressl: works

after-haproxy 2.0 opnsense
openssl: works __without__ X25519

Error with X25519:
```
unable to set SSL curves list to 'secp384r1:X25519' for bind 'internalip:443' at [/usr/local/etc/haproxy.conf:40].
```

libressl: doesnt work with any, not even openssl ec/ecparam, listed curves. If any (valid) curve is specified it refuses to work.

Error: 
```
 'bind internalip:443' : 'curves' : library does not support curve suite
```


Full-ish config

```
#
# Automatically generated configuration.
# Do not edit this file manually.
#

global
    # NOTE: Could be a security issue, but required for some feature.
    uid                         80
    gid                         80
    chroot                      /var/haproxy
    daemon
    stats                       socket /var/run/haproxy.socket level admin
    nbproc                      1
    nbthread                    1
    maxconn                     400
    tune.ssl.default-dh-param   2048
    spread-checks               2
    tune.chksize                16384
    tune.bufsize                16384
    tune.lua.maxmem             0
    log /var/run/log local0 info
    ssl-default-bind-options no-sslv3 no-tlsv10 no-tlsv11 no-tls-tickets
    ssl-default-bind-ciphers TLS13-CHACHA20-POLY1305-SHA256:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-ECDSA-AES256-SHA384:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-CHACHA20-POLY1305:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-SHA384:!aNULL:!eNULL:!EXPORT:!DES:!RC4:!MD5:!PSK:!aECDH:!EDH-DSS-DES-CBC3-SHA:!EDH-RSA-DES-CBC3-SHA:!KRB5-DES-CBC3-SHA

defaults
    log     global
    option redispatch -1
    timeout client 60s
    timeout connect 60s
    timeout server 60s
    retries 3

# autogenerated entries for ACLs

# autogenerated entries for config in backends/frontends

# autogenerated entries for stats


# Frontend: HTTP_PUB (Public HTTP Redirect/LE assign)
frontend HTTP_PUB
    bind serveripv4:80 name serveripv4:80 
    bind serveripv6:80 name serveripv6:80 
    mode http
    option http-keep-alive
    # tuning options
    timeout client 60s

    # logging options
    # ACL: find_acme_challenge
    acl acl4 path_beg -i /.well-known/acme-challenge/
    # ACL: IS_NOT_SSLacl_
    acl acl5 ssl_fc

    # ACTION: EnforceHttps
    http-request redirect scheme https code 301 unless !acl5 acl4

    # ACTION: redirect_acme_challenges
    use_backend acme_challenge_backend if acl4

# Frontend: HTTPS_PUB (Public HTTPS)
frontend HTTPS_PUB
    http-response set-header Strict-Transport-Security "max-age=15768000"
    bind internalip:443 name internalip:443 curves X25519:secp521r1:secp384r1 ssl no-sslv3 no-tlsv10 no-tlsv11 no-tls-tickets ciphers TLS_CHACHA20_POLY1305_SHA256:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-AES128-GCM-SHA256 alpn h2,http/1.1 crt-list /tmp/haproxy/ssl/certpath.certlist 
    bind serveripv4:443 name serveripv4:443 curves X25519:secp521r1:secp384r1 ssl no-sslv3 no-tlsv10 no-tlsv11 no-tls-tickets ciphers TLS_CHACHA20_POLY1305_SHA256:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-AES128-GCM-SHA256 alpn h2,http/1.1 crt-list /tmp/haproxy/ssl/certpath.certlist 
    bind serveripv6:443 name serveripv6:443 curves X25519:secp521r1:secp384r1 ssl no-sslv3 no-tlsv10 no-tlsv11 no-tls-tickets ciphers TLS_CHACHA20_POLY1305_SHA256:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-AES128-GCM-SHA256 alpn h2,http/1.1 crt-list /tmp/haproxy/ssl/certpath.certlist 
    mode http
    option http-keep-alive
    option forwardfor
    # tuning options
    timeout client 60s

    # logging options
    # ACL: myacl1
    acl acl1 hdr(host) -i domain1
    # ACL: IS_SSL
    acl acl2 ssl_fc
    # ACL: myacl2
    acl acl3 hdr(host) -i domain2
    # ACL: myacl3
    acl acl_51 hdr_reg(host) -i domain3
    # ACL: myacl4
    acl acl_52 hdr(host) -i domain4
    # ACL: myacl5
    acl acl_53 hdr(host) -i domain5


    # ACTION: DeleteXPoweredBy
    http-response del-header X-Powered-By if acl2
    # ACTION: DeleteServer
    http-response del-header Server if acl2

    # ACTION: 1_action
    use_backend 1_backend if acl1
    # ACTION: 2_action
    use_backend 2_backend if acl3
    # ACTION: 3_action
    use_backend 3_backend if acl_51
    # ACTION: 4_action
    use_backend 4_backend if acl_52
    # ACTION: 5_action
    use_backend 5_backend if acl_53
    # WARNING: pass through options below this line
    option forwardfor

# Backend: acme_challenge_backend (Added by Let's Encrypt plugin)
backend acme_challenge_backend
    # health checking is DISABLED
    mode http
    balance source
    # stickiness
    stick-table type ip size 50k expire 30m  
    stick on src
    # tuning options
    timeout connect 60s
    timeout server 60s
    http-reuse never
    server acme_challenge_host acmesrchere


backend example_https_backend
    # health checking is DISABLED
    mode http
    balance source
    # stickiness
    stick-table type ip size 50k expire 30m  
    stick on src
    # tuning options
    timeout connect 60s
    timeout server 60s
    http-reuse never
    server my_server_1 internalip2:443 ssl verify required ca-file /tmp/haproxy/ssl/capath.calist


backend example_http_backend
    # health checking is DISABLED
    mode http
    balance source
    # stickiness
    stick-table type ip size 50k expire 30m  
    stick on src
    # tuning options
    timeout connect 60s
    timeout server 60s
    http-reuse never
    server my_server_2 internalip1:80 
```

**Comment by wtarreau** [comment]
Thank you, this is exactly what I needed (because I didn't know this syntax
nor feature). I can now reproduce it, but it doesn't seem to be a regression
at all, as I get the same error in 1.8, 1.9, 2.0 and 2.1:

   [ALERT] 323/044308 (6748) : Proxy 'foo': unable to set SSL curves list to 'X25519:secp521r1:secp384r1' for bind ':4443' at [curves.cfg:2].

And I confirm that removing "X25519" from the list works. Note that this is
with openssl 1.0.2. I'll have to try on another machine with a more recent
openssl version.

**Comment by wtarreau** [comment]
OK now tested on a machine equiped with openssl 1.1.1 and I see that it <b>does</b> work fine for 1.8 to 2.1. Just rebuilding the same code with openssl 1.0.2 then fails. Thus I can affirm that it is unrelated to the haproxy version and is exclusively related to the SSL library in use. Now regarding libressl support, it's only in best effort mode since libressl only includes a subset of openssl's API. If you think that this feature used to work in a previous haproxy version with the *exact same* libressl version and does not work anymore with a more recent haproxy version, then it could be considered as a regression (in which case all these versions would be needed). But at this point I'm unable to verify this and my tests with openssl do not seem to support this theory.

Now regarding openssl support, since your executable above was built against libressl, are you certain that you tested with openssl and more specifically an up-to-date version ? I suspect that you could inadvertently have switched to an older version during your different tests.

**Comment by chipitsine** [comment]
building custom openssl (and linking haproxy to it) may be done like that

openssl itself

```
export PREFIX=/opt/openssl

git clone --depth 1 -b OpenSSL_1_1_1-stable https://github.com/openssl/openssl.git
cd openssl
./config shared -d --prefix=${PREFIX} --openssldir=${PREFIX} -Wl,-rpath=${PREFIX}/lib
 make all 
make install 
cd ..
```

haproxy

```
export PREFIX=/opt/openssl
export LDFLAGS="-L${PREFIX}/lib -Wl,-rpath,${PREFIX}/lib"
.... make as usual
```

please note `rpath` and `PREFIX`

we do use custom openssl builds in our CI (it uses LD_LIBRARY_PATH which is worse than rpath)
maybe it should be documented somewhere (the way how to build and link to custom openssl)

**Comment by chipitsine** [comment]
here's LibreSSL 3.0.2 build

https://travis-ci.com/chipitsine/haproxy/jobs/258516053#L774

seems, LibreSSL supports only sec256 curves so far

**Comment by chipitsine** [comment]
he-he, sec256r1 (the most secure one) is not supported

**Comment by wtarreau** [comment]
Ilya, I think that the various ways to build for a specific openssl version should in fact be documented in the wiki, this would be a very good use of it ;-)

**Comment by waffshappen** [comment]
> OK now tested on a machine equiped with openssl 1.1.1 and I see that it does work fine for 1.8 to 2.1. Just rebuilding the same code with openssl 1.0.2 then fails. Thus I can affirm that it is unrelated to the haproxy version and is exclusively related to the SSL library in use. Now regarding libressl support, it's only in best effort mode since libressl only includes a subset of openssl's API. If you think that this feature used to work in a previous haproxy version with the _exact same_ libressl version and does not work anymore with a more recent haproxy version, then it could be considered as a regression (in which case all these versions would be needed). But at this point I'm unable to verify this and my tests with openssl do not seem to support this theory.
> 
> Now regarding openssl support, since your executable above was built against libressl, are you certain that you tested with openssl and more specifically an up-to-date version ? I suspect that you could inadvertently have switched to an older version during your different tests.

The specific versions running are:
openssl	1.0.2t
haproxy	2.0.8

And i am 100% certain that it used to work, i even have screenshots of **X25519** (and **secp512r1**) being used as cipher, with libressl, on an older version.
Openssl doesnt handle the cipher `ECDHE-ECDSA-CHACHA20-POLY1305` in the opnsense setup, so having libressl work again would be the ideal case. I'll ask the plugin maintainer about the above build changes if they'd have any effect or are in place already.

So could it be that the case just is haproxy 2.x with openssl <1.1 just doesnt work with x25519 anymore, for now? (and libressl 2.x being completely broken for "curves").

**Comment by wtarreau** [comment]
Maybe the version in your screenshot was not 1.0.2 or maybe it included some backports. As I said, here I'm seeing this:
- haproxy 1.8, 1.9, 2.0, 2.1 with openssl 1.0.2: fail
- haproxy 1.8, 1.9, 2.0, 2.1 with openssl 1.1.1: work
So quite obviouly the only thing that varies here is the openssl version, thus it does not seem realistic that a previous version was working with the exact same setup just changing haproxy.

Regarding the cipher you're interested in, apparently it's supported in 1.1.1 which is the currently supported openssl version, I don't know if it's available in your distro:
   https://www.openssl.org/docs/man1.1.1/man1/ciphers.html

**Comment by waffshappen** [comment]
> Maybe the version in your screenshot was not 1.0.2 or maybe it included some backports. As I said, here I'm seeing this:
> 
> * haproxy 1.8, 1.9, 2.0, 2.1 with openssl 1.0.2: fail
> * haproxy 1.8, 1.9, 2.0, 2.1 with openssl 1.1.1: work
>   So quite obviouly the only thing that varies here is the openssl version, thus it does not seem realistic that a previous version was working with the exact same setup just changing haproxy.

But given its in the 1.8 docs shouldnt it also work with the older openssl versions, as i recall it?

> 
> Regarding the cipher you're interested in, apparently it's supported in 1.1.1 which is the currently supported openssl version, I don't know if it's available in your distro:
> https://www.openssl.org/docs/man1.1.1/man1/ciphers.html

Not yet, and likely not until the next big release at minimum ( Its https://opnsense.org/ )

This has already been handled since libressl 2.7, i found a screenshot i did from when it was being introduced:

![libressl](https://user-images.githubusercontent.com/44290023/69242049-55c70280-0ba0-11ea-9d24-2ca81bc2abcd.jpg)


And the handshake simulation from qualys: (old, now using only ecc certs)
![qualys_handshakes](https://user-images.githubusercontent.com/44290023/69242150-87d86480-0ba0-11ea-8e71-cda0cd9509dc.jpg)

Again, i know for a fact it has been working with older versions (curves), -with libressl however-. And it has been in the docs since 1.8, atleast according to https://cbonte.github.io/haproxy-dconv/1.8/configuration.html - since which i use it with curves x25519 set, and browsers also using it.

However, it could be that opnsense patched openssl for this? And i'm not sure they'll roll out openssl 1.1 before their next major release, so thats still several months away. ^^'


So question is - can this be handled for libressl 2.x again, whatever made it work in older versions? (unless opnsense patched it in?)

Or do i have to switch to, and wait for, openssl 1.1 to land everywhere?

**Comment by wtarreau** [comment]
I have no idea. libressl is everything but a reasonably maintained project :-(  First they broke everything by claiming they were openssl version 2.0. Then they do not seem to support yet the up-to-date parts of openssl. And by being maintained for <b>one year only</b>, it's unsuitable for shipping in any maintained project or distro. Mind you that the version that was available when haproxy 1.9 was released is not supported anymore! It's hard to make any comparison between versions this way nor to guess if any future version will work or break. Do you know what version you were running with this test ? It might make sense to try to reproduce this (same haproxy version, same libressl version). Are you really certain it was libressl ? I'm asking because the person having added this "curves" directive also heavily uses boringssl, so maybe it was this one you tested on. Just guessing :-/

**Comment by waffshappen** [comment]
> And by being maintained for one year only, it's unsuitable for shipping in any maintained project or distro. 

Fine by me with opnsense's and fedora's half-year release cycle, but even then  openssl 1.0 doesnt support x25519 it seems sadly (without patches*).

>Mind you that the version that was available when haproxy 1.9 was released is not supported anymore! It's hard to make any comparison between versions this way nor to guess if any future version will work or break. Do you know what version you were running with this test? 

I can only guess, but it worked absolutely flawless until haproxy 2.x was introduced from opnsense upstream. And they dont know either, and suggested a regression in 2.x so thats why i am asking here.
 
So likely from 1.8 ha+2.7lssl until 1.9 latest and the latest libressl 2.x

>It might make sense to try to reproduce this (same haproxy version, same libressl version). Are you really certain it was libressl ?

Yes, and still is.

> I'm asking because the person having added this "curves" directive also heavily uses boringssl, so maybe it was this one you tested on. Just guessing :-/

boringssl isnt shipped by opnsense, only open/libressl. And thats what haproxy was compiled against since forever, as you can see above in the debug info.

Sorry i cant give any insight into specific patches or compilation from opnsense, as i am merely a mortal user of it. I could look through the repo if they add any patches however. 
If you want to ask specific things related to haproxy and version etc. look/ask in the linked bug report, the maintainer of the haproxy extension for opnsense is available in that repo.

**Comment by lukastribus** [comment]
Since 9a1ab081607984bf3bfd2a7f8bf411ad0c19cd0f (`CLEANUP: ssl-sock: use HA_OPENSSL_VERSION_NUMBER instead of OPENSSL_VERSION_NUMBER`) - which is in haproxy 2.0 - we restrict LibreSSL to the OpenSSL 1.0.1 API, to avoid breaking LibreSSL every minute.

We [set HA_OPENSSL_VERSION_NUMBER to 0x1000107fL if LibreSSL is detected](https://github.com/haproxy/haproxy/blob/47479eb0e7d2f559836f02e9096308844ec2632f/include/common/openssl-compat.h#L27) and only allow curves to be configured if [HA_OPENSSL_VERSION_NUMBER  is at least 0x1000200fL](https://github.com/haproxy/haproxy/blob/47479eb0e7d2f559836f02e9096308844ec2632f/src/ssl_sock.c#L8476).


Can you try the following patch:

```
diff --git a/src/ssl_sock.c b/src/ssl_sock.c
index 6513760..4024248 100644
--- a/src/ssl_sock.c
+++ b/src/ssl_sock.c
@@ -5030,7 +5030,7 @@ int ssl_sock_prepare_ctx(struct bind_conf *bind_conf, struct ssl_bind_conf *ssl_
        if (ssl_conf_cur)
                SSL_CTX_set_alpn_select_cb(ctx, ssl_sock_advertise_alpn_protos, ssl_conf_cur);
 #endif
-#if HA_OPENSSL_VERSION_NUMBER >= 0x1000200fL
+#if ((HA_OPENSSL_VERSION_NUMBER >= 0x1000200fL) || defined(LIBRESSL_VERSION_NUMBER))
        conf_curves = (ssl_conf && ssl_conf->curves) ? ssl_conf->curves : bind_conf->ssl_conf.curves;
        if (conf_curves) {
                if (!SSL_CTX_set1_curves_list(ctx, conf_curves)) {
@@ -8473,7 +8473,7 @@ static int bind_parse_crl_file(char **args, int cur_arg, struct proxy *px, struc
 /* parse the "curves" bind keyword keyword */
 static int ssl_bind_parse_curves(char **args, int cur_arg, struct proxy *px, struct ssl_bind_conf *conf, char **err)
 {
-#if HA_OPENSSL_VERSION_NUMBER >= 0x1000200fL
+#if ((HA_OPENSSL_VERSION_NUMBER >= 0x1000200fL) || defined(LIBRESSL_VERSION_NUMBER))
        if (!*args[cur_arg + 1]) {
                if (err)
                        memprintf(err, "'%s' : missing curve suite", args[cur_arg]);

```

**Comment by wtarreau** [comment]
@lukastribus this is an extremely likely candidate indeed. However it doesn't explain why @waffshappen thinks it used to work with openssl 1.0.2 and no more, while for me 1.0.2 doesn't work and already didn't in haproxy 1.8. But I'm fine if we take openssl out of the equation and only focus on libressl if that's the only one that's certain there.

**Comment by lukastribus** [comment]
The patch above was adjusting the wrong place, I've fixed it now twice. Please use the latest patch from 09:17 UTC which changes two different lines.

I agree the openssl question is unclear at this point.

**Comment by wtarreau** [comment]
Lukas, I should issue another 2.0 tomorrow, if you think your patch is OK (and at first glance it looks so), do not hesitate to send it to me with a commit message so that I integrate it into 2.0.10.

**Comment by lukastribus** [comment]
I tested it quickly against LibreSSL 2.9.2 and can confirm that the patch fixes this issue.

I submitted it on the ML:
https://www.mail-archive.com/haproxy@formilux.org/msg35482.html

Should be backported to 2.0.

**Comment by wtarreau** [comment]
I've queued it for 2.0 now, thank you very much for your tests Lukas!

**Comment by lukastribus** [comment]
The fix has been committed and backported to 2.0, which will be released as `2.0.10`.


@waffshappen regarding the OpenSSL issues I think there is some confusion here:


> `ssl-default-bind-ciphers TLS13-CHACHA20-POLY1305-SHA256`

There is no such cipher. The "TLS13" prefixed ciphers are a relict from the OpenSSL TLSv1.3 development and never made it to any stable release.


> it would it doesnt handle `TLS_CHACHA20_POLY1305_SHA256` on opnsense for me as cipher.

That's a TLSv1.3 cipher and TLSv1.3 is only supported in OpenSSL 1.1.1. Not in older release and not in LibreSSL.


> Openssl doesnt handle the cipher `ECDHE-ECDSA-CHACHA20-POLY1305` in the opnsense setup

This requires at least OpenSSL 1.1.0 as well as a ECC certificate. OpenSSL 1.0.2 does not support Chacha20/Poly1305.


> In haproxy, with both openssl and libressl in previous versions the following in "advanced options: bind options pass-through" used to work:
> `curves X25519:secp521r1:secp384r1`

Not with OpenSSL 1.0.2, OpenSSL 1.0.2 simply does not support X25519:


```
lukas@dev:~$ openssl version
OpenSSL 1.0.2g  1 Mar 2016
lukas@dev:~$ openssl s_client -curves secp384r1 -connect 127.0.0.1:443
connect: Connection refused
connect:errno=111
lukas@dev:~$ openssl s_client -curves X25519 -connect 127.0.0.1:443
Error with command: "-curves X25519"
usage: s_client args
[...]
```


OpenSSL 1.1.0 appears to work:

```
lukas@dev:~$ ~/libsslbuild110k/bin/openssl version
OpenSSL 1.1.0k  28 May 2019
lukas@dev:~$ ~/libsslbuild110k/bin/openssl s_client -curves secp384r1 -connect 127.0.0.1:443
140403346040576:error:0200206F:system library:connect:Connection refused:crypto/bio/b_sock2.c:108:
140403346040576:error:2008A067:BIO routines:BIO_connect:connect error:crypto/bio/b_sock2.c:109:
connect:errno=111
lukas@dev:~$ ~/libsslbuild110k/bin/openssl s_client -curves X25519 -connect 127.0.0.1:443
140335689447168:error:0200206F:system library:connect:Connection refused:crypto/bio/b_sock2.c:108:
140335689447168:error:2008A067:BIO routines:BIO_connect:connect error:crypto/bio/b_sock2.c:109:
connect:errno=111
lukas@dev:~$ ~/libsslbuild110k/bin/openssl s_client -curves X255XX -connect 127.0.0.1:443
Error with command: "-curves X255XX"
lukas@dev:~$
```


So I think OpenSSL 1.0.2 never worked with X25519 in the first place.
